tests/lean/run/2161.lean

```diff
@@ -0,0 +1,50 @@
+structure Foo where num : Nat deriving DecidableEq
+
+namespace Foo
+
+instance : OfNat Foo n := ⟨⟨n⟩⟩
+
+/-! # Example 1 -/
+
+@[irreducible] def mul (a b : Foo) : Foo :=
+  let d := Nat.gcd a.num 1
+  ⟨(a.num.div d) * (b.num.div d)⟩
+
+-- should fail fast; exact heartbeat count at time of writing is 31
+set_option maxHeartbeats 310
+/--
+error: tactic 'decide' failed for proposition
+  ((mul 4 1).mul 1).mul 1 = 4
+since its 'Decidable' instance
+  instDecidableEqFoo (((mul 4 1).mul 1).mul 1) 4
+did not reduce to 'isTrue' or 'isFalse'.
+
+After unfolding the instances 'decEqFoo✝', 'instDecidableEqFoo', 'instDecidableEqNat' and
+'Nat.decEq', reduction got stuck at the 'Decidable' instance
+  match h : (((mul 4 1).mul 1).mul 1).num.beq 4 with
+  | true => isTrue ⋯
+  | false => isFalse ⋯
```

Comment:
Possible future addition: add another "Hint" covering this situation (one of the definitions is marked as irreducible)?

Reply:
It would be nice to have a more general system to diagnose reduction issues! Right now, I'm a little worried that this reduction blamer might mis-blame `match` discriminants. At least every Decidable instance that appears *should* be reducible, so it should be harmless. With arbitrary discriminants, there's more risk of being wrong, unless perhaps we report each discriminant that doesn't reduce to a constructor.

(Somewhat related, today [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Should.20wfrec.20defs.20require.20an.20.40.5Birreducible.5D.20annotation.3F/near/450769240) I was thinking about this problem for diagnosing type mismatches. It's a bit different, since it's for isDefEq, but isDefEq depends on whnf.)